src/stridelayout.jl

```diff
@@ -0,0 +1,260 @@
+    n == N || return nothing
+    ranktup = Expr(:tuple); append!(ranktup.args, rank_new) # dynamic splats bad
+    Expr(:call, Expr(:curly, :StrideRank, ranktup))
+end
+stride_rank(x, i) = stride_rank(x)[i]
+
+struct DenseDims{D} end
+Base.@pure DenseDims(D::NTuple{<:Any,Bool}) = DenseDims{D}()
+@inline Base.getindex(::DenseDims{D}, i::Integer) where {D} = D[i]
+@inline Base.getindex(::DenseDims{D}, ::Val{I}) where {D,I} = DenseDims{permute(D, I)}()
+"""
+dense_dims(::Type{T}) -> NTuple{N,Bool}
+
+Returns a tuple of indicators for whether each axis is dense.
+An axis `i` of array `A` is dense if `stride(A, i) * size(A, i) == stride(A, j)` where `stride_rank(A)[i] + 1 == stride_rank(A)[j]`.
+"""
+dense_dims(x) = dense_dims(typeof(x))
+dense_dims(::Type) = nothing
+dense_dims(::Type{Array{T,N}}) where {T,N} = DenseDims{ntuple(_ -> true, Static{N}())}()
+dense_dims(::Type{<:Tuple}) = DenseDims{(true,)}()
+function dense_dims(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}) where {T,A<:AbstractMatrix{T}}
+    dense = dense_dims(A)
+    isnothing(dense) ? nothing : dense[Val{(2,1)}()]
+end
+function dense_dims(::Type{<:PermutedDimsArray{T,N,I1,I2,A}}) where {T,N,I1,I2,A<:AbstractArray{T,N}}
+    dense = dense_dims(A)
+    isnothing(dense) ? nothing : dense[Val{I1}()]
+end
+function dense_dims(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    _dense_dims(S, dense_dims(A), stride_rank(A))
+end
+_dense_dims(::Any, ::Any) = nothing
+@generated function _dense_dims(::Type{S}, ::DenseDims{D}, ::StrideRank{R}) where {D,R,N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    still_dense = true
+    sp = rank_to_sortperm(R)
+    densev = Vector{Bool}(undef, NP)
+    for np in 1:NP
+        spₙ = sp[np]
+        still_dense &= D[spₙ]
+        densev[spₙ] = still_dense
+        # a dim not being complete makes later dims not dense
+        still_dense && (still_dense = (I.parameters[spₙ] <: Base.Slice)::Bool)
+    end
+    dense_tup = Expr(:tuple)
+    for np in 1:NP
+        spₙ = sp[np]
+        if I.parameters[np] <: Base.Slice
+            push!(dense_tup.args, densev[np])
+        elseif I.parameters[np] <: AbstractUnitRange
+            push!(dense_tup.args, densev[np])
+        end
+    end
+    # If n != N, then an axis was indexed by something other than an integer or `AbstractUnitRange`, so we return `nothing`
+    length(dense_tup.args) == N ? Expr(:call, Expr(:curly, :DenseDims, dense_tup)) : nothing
+end
+
+permute(t::NTuple{N}, I::NTuple{N,Int}) where {N} = ntuple(n -> t[I[n]], Static{N}())
+@generated function permute(t::Tuple{Vararg{Any,N}}, ::Val{I}) where {N,I}
+    t = Expr(:tuple)
+    foreach(i -> push!(t.args, Expr(:ref, :t, i)), I)
+    Expr(:block, Expr(:meta, :inline), t)
+end
+
+sdsize(::Any) = nothing
+sdstrides(::Any) = nothing
+sdoffsets(::Any) = (Static{1}(),) # Assume arbitrary Julia data structures use 1-based indexing by default.
```

Comment:
Offset indexing should have `indices` that are offset so you could probably just do

```julia
function sdoffsets(x, i)
    inds = indices(x, i)
    start = known_first(inds)
    if start === nothing
        return first(inds) - 1
    else
        return start - Static(1)
    end
end

sdoffsets(x) = ntuple(i -> sdoffsets(x, i), Val(ndims(x)))
```

This way any offset indexing is generically supported if `indices` is properly defined.

Reply:
I do want this to correspond to "X-based indexing", so that a 1-based index Array has a value of "1".
This allows, [for example](https://github.com/chriselrod/VectorizationBase.jl/blob/316209a3bb761fa09106b143586a4f174990903c/src/strided_pointers/stridedpointers.jl#L103):
```julia
tdot(map(-, i, sdoffsets(ptr)), strides(ptr), contiguous_axis_indicator(ptr))
```
to be used for calculating the linear index, rather than needing to do this offsetting here.

I'll have to clarify in the documentation then that offset means "offset from 0" and not "offset from 1".

Additionally, `known_first` currently doesn't return a `Static`.
I would be in favor of making the `known_*` functions return `Static`s, although that change would be breaking.
Do you agree?

I just added:
```julia
@inline function sdoffsets(x, i)
    inds = indices(x, i)
    start = known_first(x)
    isnothing(start) ? first(inds) : Static(start)
end
sdoffsets(A::AbstractArray{<:Any,N}) where {N} = ntuple(n -> sdoffsets(A, n), Val{N}())
```
I'll add tests tonight.

Reply:
> I'll have to clarify in the documentation then that offset means "offset from 0" and not "offset from 1".

Makes sense.

> I would be in favor of making the known_* functions return Statics, although that change would be breaking.

I'm on the fence about this one. Hopefully, it wouldn't be too breaking of a change since it should still be functionally equivalent.